beacon-chain/sync/rpc_data_column_sidecars_by_range.go

```diff
@@ -0,0 +1,191 @@
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	// Ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	batcher, err := newBlockRangeBatcher(rp, s.cfg.beaconDB, s.rateLimiter, s.cfg.chain.IsCanonical, ticker)
+	if err != nil {
+		log.WithError(err).Info("error in DataColumnSidecarsByRange batch")
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	var batch blockBatch
+	wQuota := params.BeaconConfig().MaxRequestDataColumnSidecars
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
+		batchStart := time.Now()
+		wQuota, err = s.streamBlobBatch(ctx, batch, wQuota, stream)
+		rpcBlobsByRangeResponseLatency.Observe(float64(time.Since(batchStart).Milliseconds()))
+		if err != nil {
+			return err
+		}
+		// once we have written MAX_REQUEST_BLOB_SIDECARS, we're done serving the request
+		if wQuota == 0 {
+			break
+		}
+	}
+	if err := batch.error(); err != nil {
+		log.WithError(err).Debug("error in DataColumnSidecarsByRange batch")
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	closeStream(stream, log)
+	return nil
+}
+
+// Set the count limit to the number of blobs in a batch.
+func columnBatchLimit() uint64 {
+	return uint64(flags.Get().BlockBatchLimit) / fieldparams.MaxBlobsPerBlock
+}
+
+// TODO: Generalize between data columns and blobs, while the validation parameters used are different they
+// are the same value in the config. Can this be safely abstracted ?
+func validateDataColumnsByRange(r *pb.DataColumnSidecarsByRangeRequest, current primitives.Slot) (rangeParams, error) {
+	if r.Count == 0 {
+		return rangeParams{}, errors.Wrap(p2ptypes.ErrInvalidRequest, "invalid request Count parameter")
+	}
+	rp := rangeParams{
+		start: r.StartSlot,
+		size:  r.Count,
+	}
+	// Peers may overshoot the current slot when in initial sync, so we don't want to penalize them by treating the
+	// request as an error. So instead we return a set of params that acts as a noop.
+	if rp.start > current {
+		return rangeParams{start: current, end: current, size: 0}, nil
+	}
+
+	var err error
+	rp.end, err = rp.start.SafeAdd(rp.size - 1)
+	if err != nil {
+		return rangeParams{}, errors.Wrap(p2ptypes.ErrInvalidRequest, "overflow start + count -1")
+	}
+
+	maxRequest := params.MaxRequestBlock(slots.ToEpoch(current))
+	// Allow some wiggle room, up to double the MaxRequestBlocks past the current slot,
+	// to give nodes syncing close to the head of the chain some margin for error.
+	maxStart, err := current.SafeAdd(maxRequest * 2)
```

Comment:
I don't really understand this. Why this would be ever the case?

Reply:
The caller knows the current slot. So why would it send a request with a `start > currentSlot`?